File: HotpotLayout/HotpotLayout/ViewController.m
//
//  ViewController.m
//  HotpotLayout
//
//  Created by meterscao on 16/4/20.
//  Copyright © 2016年 meters. All rights reserved.
//

#import "ViewController.h"
#import "Hotpot.h"

@interface ViewController ()
@property(nonatomic)UIView *redView;
@property(nonatomic)UIView *blueView;;
@property(nonatomic)UIView *blackView;
@property(nonatomic)UIView *orangeView;
@property(nonatomic)UILabel *nameLabel;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.redView = [[UIView alloc]init];
    self.blueView = [[UIView alloc]init];
    self.blackView = [[UIView alloc]init];
    self.orangeView = [[UIView alloc]init];
    
    self.redView.layer.cornerRadius = 4.f;
    self.redView.backgroundColor  = [[UIColor redColor]colorWithAlphaComponent:0.7f];
    self.redView.layer.masksToBounds = YES;
    [self.view addSubview:self.redView];
    
    self.blueView.backgroundColor = [[UIColor blueColor]colorWithAlphaComponent:0.8f];
    self.blueView.layer.cornerRadius = 4.f;
    self.blueView.layer.masksToBounds = YES;
    [self.view addSubview:self.blueView];
    
    self.blackView.backgroundColor = [[UIColor blackColor]colorWithAlphaComponent:0.8f];
    self.blackView.layer.cornerRadius = 4.f;
    self.blackView.layer.masksToBounds = YES;
    [self.view addSubview:self.blackView];
    
    self.orangeView.backgroundColor = [[UIColor orangeColor]colorWithAlphaComponent:0.8f];
    self.orangeView.layer.cornerRadius = 20.f;
    self.orangeView.layer.masksToBounds = YES;
    [self.view addSubview:self.orangeView];
    
    
    self.nameLabel = [[UILabel alloc]init];
    self.nameLabel.text = @"I LOVE HOTPOT";
    self.nameLabel.textAlignment = NSTextAlignmentCenter;
    self.nameLabel.backgroundColor = [[UIColor purpleColor] colorWithAlphaComponent:0.7f];
    self.nameLabel.layer.cornerRadius = 4.f;
    self.nameLabel.layer.masksToBounds = YES;
    self.nameLabel.textColor = [UIColor whiteColor];
    self.nameLabel.font = [UIFont systemFontOfSize:18.f];
    [self.view addSubview:self.nameLabel];
    
}



-(void)viewDidLayoutSubviews{
    
    [super viewDidLayoutSubviews];
    
    [self.redView frameLayout:^(HotpotFrameLayout *layout) {
        layout.left.equalTo(self.view.left).offset(20.f);
        layout.right.equalTo(self.view.right).offset(-20.f);
        layout.height.equalTo(40.f);
        layout.top.equalTo(self.view.top).offset(40.f);
    }];
    
    [self.blueView frameLayout:^(HotpotFrameLayout *layout) {
        layout.width.equalTo(100.f);
        layout.centerX.equalTo(self.view.centerX);
        layout.bottom.equalTo(self.view.bottom).offset(-50.f);
        layout.height.equalTo(40.f);
    }];
    
    [self.blackView frameLayout:^(HotpotFrameLayout *layout) {
        layout.height.equalTo(100.f);
        layout.centerY.equalTo(self.view.centerY);
        layout.left.equalTo(self.view.left).offset(20.f);
        layout.right.equalTo(self.blueView.left);
    }];
    
    [self.orangeView frameLayout:^(HotpotFrameLayout *layout) {
        layout.width.equalTo(40.f);
        layout.height.equalTo(40.f);
        layout.left.equalTo(self.blackView.left);
        layout.top.equalTo(self.blackView.bottom).offset(20.f);
    }];
    
    [self.nameLabel sizeToFit];
    [self.nameLabel frameLayout:^(HotpotFrameLayout *layout) {
        layout.right.equalTo(self.redView.right);
        layout.bottom.equalTo(self.blackView.top).offset(-20.f);
    }];
}



@end

File: HotpotLayout/HotpotLayout/ViewController.swift
//
//  ViewController.swift
//  HotpotLayout
//
//  Created by R0uter on 2018/4/1.
//  Copyright © 2018 R0uter. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    let redView = UIView()
    let blueView = UIView()
    let blackView =  UIView()
    let orangeView = UIView()
    let nameLabel = UILabel()
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        redView.layer.cornerRadius = 4
        redView.backgroundColor = UIColor.red.withAlphaComponent(0.7)
        redView.layer.masksToBounds = true
        view.addSubview(redView)
        
        blueView.backgroundColor = UIColor.blue.withAlphaComponent(0.8)
        blueView.layer.cornerRadius = 4
        blueView.layer.masksToBounds = true
        view.addSubview(blueView)
        
        blackView.backgroundColor = UIColor.black.withAlphaComponent(0.8)
        blackView.layer.cornerRadius = 4
        blackView.layer.masksToBounds = true
        view.addSubview(blackView)
        
        orangeView.backgroundColor = UIColor.orange.withAlphaComponent(0.8)
        orangeView.layer.cornerRadius = 20
        orangeView.layer.masksToBounds = true
        view.addSubview(orangeView)
        
        nameLabel.text = "I LOVE HOTPOT"
        nameLabel.textAlignment = .center
        nameLabel.backgroundColor = UIColor.purple.withAlphaComponent(0.7)
        nameLabel.layer.masksToBounds = true
        nameLabel.textColor = UIColor.white
        nameLabel.font = UIFont.systemFont(ofSize: 18)
        view.addSubview(nameLabel)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        redView.frameLayout { $0
            .left.equal(to: view.left).offset(20)
            .right.equal(to: view.right).offset(-20)
            .height.equal(to: 40)
            .top.equal(to: view.top).offset(view.layoutMargins.top+40)
        }
        
        blueView.frameLayout { $0
            .width.equal(to: 100)
            .centerX.equal(to: view.centerX)
            .bottom.equal(to: view.bottom).offset(view.layoutMargins.bottom-50)
            .height.equal(to: 40)
        }
        blackView.frameLayout { $0
            .height.equal(to: 100)
            .centerY.equal(to: view.centerY)
            .left.equal(to: view.left).offset(20)
            .right.equal(to: blueView.left)
        }
        
        orangeView.frameLayout { $0
            .width.equal(to: 40)
            .height.equal(to: 40)
            .left.equal(to: blackView.left).offset(100)
            .top.equal(to: blackView.bottom).offset(200)
        }
        
        nameLabel.frameLayout { $0
            .right.equal(to: redView.right)
            .bottom.equal(to: blackView.top).offset(-20)
        }
    }


}

